protocol/src/packet.rs

```diff
@@ -0,0 +1,190 @@
+use std::io::{self, Read, Write};
+
+use flate2::read::ZlibDecoder;
+use flate2::write::ZlibEncoder;
+use flate2::Compression;
+use minecraft_protocol_derive::{Decoder, Encoder};
+
+use crate::decoder::{Decoder, DecoderReadExt};
+use crate::encoder::{Encoder, EncoderWriteExt};
+use crate::error::{DecodeError, EncodeError};
+
+#[derive(Debug, Clone)]
+pub struct Packet {
+    pub id: i32,
+    pub data: Vec<u8>,
+}
+
+impl Packet {
+    pub fn encode<W: Write>(
+        &self,
+        writer: &mut W,
+        compression_threshold: Option<i32>,
+    ) -> Result<(), EncodeError> {
+        let mut buf = Vec::new();
+        let packet = RawPacket {
+            id: self.id,
+            data: self.data.clone(),
+        };
+        if let Some(threshold) = compression_threshold {
+            CompressedRawPacket { packet, threshold }.encode(&mut buf)?;
+        } else {
+            packet.encode(&mut buf)?;
+        }
+
+        writer.write_var_i32(buf.len() as i32)?;
+        writer.write_all(&buf)?;
+
+        Ok(())
+    }
+
+    pub fn decode<R: Read>(reader: &mut R, compressed: bool) -> Result<Packet, DecodeError> {
+        let len = match reader.read_var_i32() {
+            Ok(len) => len as usize,
+            Err(DecodeError::IoError { io_error })
+                if io_error.kind() == io::ErrorKind::UnexpectedEof =>
+            {
+                return Err(DecodeError::Incomplete { bytes_needed: 1 })
```

Comment:
This is not correct because read var int may vary from 1 to 4 bytes

Reply:
This is used only as an indication, so it should be read as “at least 1”. If more are needed, it will happen in the next iteration.

Reply:
Maybe it should be an `Option` so it works similar to https://docs.rs/nom/latest/nom/enum.Needed.html.